File: eel-core/src/main/kotlin/io/eels/component/csv/CsvPart.kt
package io.eels.component.csv

import com.univocity.parsers.csv.CsvParser
import io.eels.Row
import io.eels.schema.Schema
import io.eels.component.Part
import rx.Observable
import java.nio.file.Path

class CsvPart(val createParser: () -> CsvParser,
              val path: Path,
              val header: Header,
              val verifyRows: Boolean,
              val schema: Schema) : Part {

  val rowsToSkip: Int = when (header) {
    Header.FirstRow -> 1
    else -> 0
  }

  override fun data(): Observable<Row> {

    val parser = createParser()
    parser.beginParsing(path.toFile())

    val iterator = generateSequence { parser.parseNext() }.drop(rowsToSkip)

    return Observable.create { sub ->
      sub.onStart()
      iterator.forEach {
        val row = Row(schema, it.asList())
        sub.onNext(row)
      }
      if (!sub.isUnsubscribed)
        sub.onCompleted()
    }
  }
}

File: eel-core/src/main/scala/io/eels/component/csv/CsvPart.scala
package io.eels.component.csv

import java.io.InputStream

import com.sksamuel.exts.Logging
import com.sksamuel.exts.io.Using
import com.univocity.parsers.csv.CsvParser
import io.eels.datastream.Subscriber
import io.eels.schema.StructType
import io.eels.{Part, Row}
import org.apache.hadoop.fs.FileSystem

class CsvPart(createParser: () => CsvParser,
              inputFn: () => InputStream,
              header: Header,
              skipBadRows: Boolean,
              schema: StructType)
             (implicit fs: FileSystem) extends Part with Logging with Using {

  val rowsToSkip: Int = header match {
    case Header.FirstRow => 1
    case _ => 0
  }

  override def subscribe(subscriber: Subscriber[Seq[Row]]): Unit = {
    using(inputFn()) { input =>

      val parser = createParser()

      try {

        parser.beginParsing(input)

        val iterator = Iterator.continually(parser.parseNext).takeWhile(_ != null).drop(rowsToSkip).map { records =>
          Row(schema, records.toVector)
        }

        iterator.grouped(1000).foreach(subscriber.next)
        subscriber.completed()

      } catch {
        case t: Throwable => subscriber.error(t)
      } finally {
        parser.stopParsing()
        input.close()
      }
    }
  }
}